prov/efa/src/efa_rma.c

```diff
@@ -0,0 +1,399 @@
+/* SPDX-FileCopyrightText: Copyright Amazon.com, Inc. or its affiliates. All rights reserved. */
+
+#include <stdlib.h>
+#include <string.h>
+#include <ofi_mem.h>
+#include <ofi_iov.h>
+#include "efa.h"
+#include "efa_av.h"
+
+#define EFA_SETUP_IOV(iov, buf, len)         \
+	do {                                 \
+		iov.iov_base = (void *) buf; \
+		iov.iov_len = (size_t) len;  \
+	} while (0)
+
+#define EFA_SETUP_RMA_IOV(rma_iov, _addr, _len, _key) \
+	do {                                          \
+		rma_iov.addr = (uint64_t) _addr;      \
+		rma_iov.len = (size_t) _len;          \
+		rma_iov.key = (uint64_t) _key;        \
+	} while (0)
+
+#define EFA_SETUP_MSG_RMA(msg, iov, _desc, count, _addr, _rma_iov,  \
+			  _rma_iov_count, _context, _data)          \
+	do {                                                        \
+		msg.msg_iov = (const struct iovec *) iov;           \
+		msg.desc = (void **) _desc;                         \
+		msg.iov_count = (size_t) count;                     \
+		msg.addr = (fi_addr_t) _addr;                       \
+		msg.rma_iov = (const struct fi_rma_iov *) _rma_iov; \
+		msg.rma_iov_count = (size_t) _rma_iov_count;        \
+		msg.context = (void *) _context;                    \
+		msg.data = (uint32_t) _data;                        \
+	} while (0)
+
+/**
+ * @brief check whether endpoint was configured with FI_RMA capability
+ * @return -FI_EOPNOTSUPP if FI_RMA wasn't requested, 0 if it was.
+ */
+static inline int efa_rma_check_cap(struct efa_base_ep *base_ep) {
+	if ((base_ep->info->caps & FI_RMA) == FI_RMA)
+		return 0;
+	EFA_WARN_ONCE(FI_LOG_EP_DATA, "Operation requires FI_RMA capability, which was not requested.\n");
+	return -FI_EOPNOTSUPP;
+}
+
+/*
+ * efa_rma_post_read() will post a read request.
+ *
+ * Input:
+ *     base_ep: endpoint
+ *     msg: read operation information
+ *     flags: currently no flags is taken
+ *
+ * On success return 0,
+ * If read failed, return the error of read operation
+ */
+static inline ssize_t efa_rma_post_read(struct efa_base_ep *base_ep,
+					const struct fi_msg_rma *msg,
+					uint64_t flags)
+{
+	struct efa_qp *qp;
+	struct efa_mr *efa_mr;
+	struct efa_conn *conn;
+#ifndef _WIN32
+	struct ibv_sge sge_list[msg->iov_count];
+#else
+	/* MSVC compiler does not support array declarations with runtime size, so hardcode
+	 * the expected iov_limit/max_sq_sge from the lower-level efa provider.
+	 */
+	struct ibv_sge sge_list[EFA_DEV_ATTR_MAX_WR_SGE];
+#endif
+	int i, err = 0;
+
+	assert(msg->iov_count > 0 &&
+	       msg->iov_count <= base_ep->domain->info->tx_attr->iov_limit);
+	assert(msg->rma_iov_count > 0 &&
+	       msg->rma_iov_count <= base_ep->domain->info->tx_attr->rma_iov_limit);
+	assert(ofi_total_iov_len(msg->msg_iov, msg->iov_count) <=
+	       base_ep->domain->device->max_rdma_size);
+
+	qp = base_ep->qp;
+	if (!base_ep->is_wr_started) {
+		ibv_wr_start(qp->ibv_qp_ex);
+		base_ep->is_wr_started = true;
+	}
+	qp->ibv_qp_ex->wr_id = (uintptr_t)msg->context;
```

Comment:
I think a future change will be like we will enforce `FI_CONTEXT`, and if flags doesn't have `FI_COMPLETION`, we will ignore `context` and make `wr_id` as NULL. Then in the cq read, we will treat NULL wr_id as the indicator of no-completion. This can be a future PR though.